libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,690 @@
+/*
+  C++ implementation of quicktune based on original lua script
+ */
+
+// quicktune is not performance sensitive, save flash
+#pragma GCC optimize("Os")
+
+#include "AP_Quicktune.h"
+
+#if AP_QUICKTUNE_ENABLED
+
+#define UPDATE_RATE_HZ 40
+#define UPDATE_PERIOD_MS (1000U/UPDATE_RATE_HZ)
+#define STAGE_DELAY 4000
+#define PILOT_INPUT_DELAY 4000
+#define YAW_FLTE_MAX 2.0
+#define FLTD_MUL 0.5
+#define FLTT_MUL 0.5
+#define DEFAULT_SMAX 50.0
+#define OPTIONS_TWO_POSITION (1<<0)
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Vehicle/AP_Vehicle.h>
+#include <AP_RCMapper/AP_RCMapper.h>
+#include <AP_Logger/AP_Logger.h>
+#include <AP_Arming/AP_Arming.h>
+#include <GCS_MAVLink/GCS.h>
+
+const AP_Param::GroupInfo AP_Quicktune::var_info[] = {
+    // @Param: ENABLE
+    // @DisplayName: Quicktune enable
+    // @Description: Enable quicktune system
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Standard
+    AP_GROUPINFO_FLAGS("ENABLE", 1, AP_Quicktune, enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: AXES
+    // @DisplayName: Quicktune axes
+    // @Description: Axes to tune
+    // @Bitmask: 0:Roll,1:Pitch,2:Yaw
+    // @User: Standard
+    AP_GROUPINFO("AXES", 2, AP_Quicktune, axes_enabled, 7),
+
+    // @Param: DOUBLE_TIME
+    // @DisplayName: Quicktune doubling time
+    // @Description: Time to double a tuning parameter. Raise this for a slower tune.
+    // @Range: 5 20
+    // @Units: s
+    // @User: Standard
+    AP_GROUPINFO("DOUBLE_TIME", 3, AP_Quicktune, double_time, 10),
+
+    // @Param: GAIN_MARGIN
+    // @DisplayName: Quicktune gain margin
+    // @Description: Reduction in gain after oscillation detected. Raise this number to get a more conservative tune
+    // @Range: 20 80
+    // @Units: %
+    // @User: Standard
+    AP_GROUPINFO("GAIN_MARGIN", 4, AP_Quicktune, gain_margin, 60),
+
+    // @Param: OSC_SMAX
+    // @DisplayName: Quicktune oscillation rate threshold
+    // @Description: Threshold for oscillation detection. A lower value will lead to a more conservative tune.
+    // @Range: 1 10
+    // @User: Standard
+    AP_GROUPINFO("OSC_SMAX", 5, AP_Quicktune, osc_smax, 5),
+
+    // @Param: YAW_P_MAX
+    // @DisplayName: Quicktune Yaw P max
+    // @Description: Maximum value for yaw P gain
+    // @Range: 0.1 3
+    // @User: Standard
+    AP_GROUPINFO("YAW_P_MAX", 6, AP_Quicktune, yaw_p_max, 0.5),
+
+    // @Param: YAW_D_MAX
+    // @DisplayName: Quicktune Yaw D max
+    // @Description: Maximum value for yaw D gain
+    // @Range: 0.001 1
+    // @User: Standard
+    AP_GROUPINFO("YAW_D_MAX", 7, AP_Quicktune, yaw_d_max, 0.01),
+
+    // @Param: RP_PI_RATIO
```

Comment:
I would note that this is a ratio and else where in this PR there are strong arguments for percentages. Should they at least all be the same?

Reply:
we could do it as a percentage, but in common parlance when we talk to users we normally discuss this as a ratio, unlike backoff which we normally discuss as a percentage. 
Looking in our wiki at the moment we have a few places where we discuss P to I as a ratio (eg. qhover-mode tuning)
would you actually prefer this one as a percentage?
I certainly don't think that all parameters that can be a percentage should be a percentage. You have to think about which one best communicates the meaning to the user

Reply:
I don't really have any strong feelings here. I think we should try to be as consistent as we can. Consistency with the wiki is just as valid as consistency with the other params added here. At least the pram name is explicitly "RATIO", so hopefully that will minimise confusion. 

I just wanted to highlight that this was a ratio having read the arguments for percentage made in other comments.